liwords-ui/src/gameroom/board.tsx

```diff
@@ -26,13 +26,159 @@ type Props = {
+    top: 0,
+    width: 1,
+    height: 1,
+  });
+  const resizeFunc = React.useCallback(() => {
+    const boardElt = boardEltRef.current;
+    if (boardElt) {
+      const { left, top, width, height } = boardElt.getBoundingClientRect();
+      setBoardSize({
+        left,
+        top,
+        width: Math.max(1, width),
+        height: Math.max(1, height),
+      });
+    }
+  }, []);
+  const boardRef = React.useCallback(
+    (elt) => {
+      boardEltRef.current = elt;
+      resizeFunc();
+    },
+    [resizeFunc]
+  );
+  React.useEffect(() => {
+    window.addEventListener('resize', resizeFunc);
+    return () => window.removeEventListener('resize', resizeFunc);
+  }, [resizeFunc]);
+  const getXY = React.useCallback(
+    (evt: React.MouseEvent): { x: number; y: number } => {
+      const x = Math.max(
+        0,
+        Math.min(1, (evt.clientX - boardSize.left) / boardSize.width)
+      );
+      const y = Math.max(
+        0,
+        Math.min(1, (evt.clientY - boardSize.top) / boardSize.height)
+      );
+      return { x, y };
+    },
+    [boardSize]
+  );
+  const [picture, setPicture] = React.useState<{
+    drawing: boolean;
+    picture: Array<Array<{ x: number; y: number }>>;
+  }>({ drawing: false, picture: [] });
+  const hasPicture = picture.picture.length > 0;
+  const handleContextMenu = React.useCallback(
+    (evt: React.MouseEvent) => {
+      if (!evt.shiftKey) {
+        // Draw when not holding shift.
+        evt.preventDefault();
+      } else if (hasPicture) {
+        // Shift+RightClick clears drawing.
+        setPicture((pic) => ({ ...pic, drawing: false, picture: [] }));
+        evt.preventDefault();
+      } else {
+        // Shift+RightClick accesses context menu if no drawing.
+      }
+    },
+    [hasPicture]
+  );
+  const handleMouseDown = React.useCallback(
+    (evt: React.MouseEvent) => {
+      if (evt.button === 2 && !evt.shiftKey) {
+        const newXY = getXY(evt);
+        setPicture((pic) => {
+          pic.picture.push([newXY]); // mutate
+          return { ...pic, drawing: true }; // shallow clone for performance
+        });
+      }
+    },
+    [getXY]
+  );
+  const handleMouseUp = React.useCallback((evt: React.MouseEvent) => {
+    setPicture((pic) => {
+      if (!pic.drawing) return pic;
+      // Right-click this many times to clear drawing.
+      const howMany = 3;
+      if (pic.picture.length >= howMany) {
+        let i = 0;
+        for (; i < howMany; ++i) {
+          if (
+            !(
+              pic.picture[pic.picture.length - (i + 1)].length < 2 &&
+              pic.picture[pic.picture.length - (i + 1)][0].x ===
+                pic.picture[pic.picture.length - 1][0].x &&
+              pic.picture[pic.picture.length - (i + 1)][0].y ===
+                pic.picture[pic.picture.length - 1][0].y
+            )
+          )
+            break;
+        }
+        if (i === howMany) {
+          return { ...pic, drawing: false, picture: [] };
```

Comment:
How does this work? Is there some criteria for how fast the three right-clicks must be?

Reply:
No time limit but they must be at the exact same spot.
Alternatively Shift+RightClick also clears if there's a drawing (otherwise I'm letting it open the context menu)

Reply:
This is neat, but pretty clunky UX to draw with the right mouse pressed. Let's move it to its own component, not call in in yet, and work on it some more when we get time. I'd like it to eventually be something that has to be turned on first (maybe with like a key command), because it will definitely confuse the average user if they right click by accident and can't figure out how to remove.)

Reply:
Refactored accordingly. When we've figured out how we want to enable it, it's just a `setEnableDrawing(true);` away.